packages/teleterm/src/services/config/configService.ts

```diff
@@ -1,39 +1,85 @@
-import { merge } from 'lodash';
+import { merge, set as lodashSet, get as lodashGet, cloneDeep } from 'lodash';
+
+import { FileStorage } from 'teleterm/services/fileStorage';
 
 import { keyboardShortcutsConfigProvider } from './providers/keyboardShortcutsConfigProvider';
 import { appearanceConfigProvider } from './providers/appearanceConfigProvider';
-import { Config, ConfigService, ConfigServiceProvider } from './types';
+import { usageMetricsConfigProvider } from './providers/usageMetricsConfigProvider';
+import {
+  Config,
+  ConfigGetter,
+  ConfigRestorer,
+  ConfigService,
+  ConfigServiceProvider,
+  ConfigSetter,
+} from './types';
 
 type ConfigServiceProviders = {
   readonly [Property in keyof Config]: ConfigServiceProvider<Config[Property]>;
 };
 
 export class ConfigServiceImpl implements ConfigService {
-  private config: Config;
+  private readonly defaultConfig: Config;
+  private userConfig: Partial<Config>;
+  private mergedConfig: Config;
+  private version = '1';
   private configProviders: ConfigServiceProviders = {
     keyboardShortcuts: keyboardShortcutsConfigProvider,
     appearance: appearanceConfigProvider,
+    usageMetrics: usageMetricsConfigProvider,
   };
 
-  constructor() {
-    this.createDefaultConfig();
+  constructor(private appConfigFileStorage: FileStorage) {
+    this.defaultConfig = this.getDefaultConfig();
+    this.userConfig = this.readUserConfig();
+    this.setConfigVersion();
+    this.produceMergedConfig();
   }
 
-  get(): Config {
-    return this.config;
-  }
+  get: ConfigGetter = path => {
+    return {
+      value: lodashGet(this.mergedConfig, path),
+      metadata: { isStored: !!lodashGet(this.userConfig, path) },
+    };
+  };
 
-  update(newConfig: Partial<Config>): void {
-    this.config = merge(this.config, newConfig);
-  }
+  set: ConfigSetter = (path, value) => {
```

Comment:
At the beginning I thought it will be easier to keep it nested, because our config structure is nested already, so no major changes will be needed. 
But now I see the drawbacks of the nested structure: 
- we need to use Lodash getters/setters, 
- we need to merge the configs, which is not that simple. 

In case of the flat structure we only need to answer a simple question: "do we have that property stored in a file?" If yes - use it, if no - take the default.  

So now I think we should go with the flat structure.

I'm only wondering how to combine it with the current config providers, I like that different parts of config are separated in different files. I see two options here:
- add a top level key to each entry (`keyboardShortcuts.tab-1`, `keyboardShortcuts.tab-2`), but it is a lot of repetition 
- flatten the config in `configService`, but it requires more work to keep it typesafe.

I'm trying the second option, I will see how it works.

Reply:
I'd also check how those two projects approach this, it seems that the structure might be "semi-flat", that is the keys in the config are flag but the actual config used by the app is not, idk.

Reply:
I will check that, but it seems to me that we should only work on flat configs. 
Let's say I have a following config type:
```
appearance: {
  fonts: {
   sansSerif: string
  }
}
```

In the code you can then change the config by: `set('appearance.fonts.sansSerif', 'Arial')`  or `set('appearance.fonts', {sansSerif: 'Arial'})`. 
Only the first option should be allowed; we can't save the second one as a flat config - it mixes property name with the value.

By defining properties this way: 
```
appearance.fonts.sansSerif: string;
terminal.customEnv: object;
```
you can easily know what is a key and what is a value.

Reply:
Right, but we still need to accept scalar values. The configs above seem to be fine with you providing a map for `terminal.integrated.env.osx` or an array for `coc.preferences.formatOnSaveFiletypes`. So essentially you can provide values only for the "leaves" of the config – but yeah that's something you'd need to check.